jobs/MeterInstanceJob.js

```diff
@@ -0,0 +1,145 @@
+      resourceType: CONST.APISERVER.RESOURCE_TYPES.SFEVENT,
+      query: {
+        labelSelector: `state in (${CONST.METER_STATE.TO_BE_METERED},${CONST.METER_STATE.FAILED})`
+      }
+    };
+    return apiServerClient.getResources(options);
+  }
+
+  /* jshint ignore:start */
+  static async meter(events) {
+    try {
+      logger.info(`Number of events to be metered in this run - ${events.length}`);
+      // Adding this comment as we are transitioning to async/await
+      // Note: The below Promise is not bluebird promise
+      const resultArray = await Promise.all(_.map(events, async (event) =>
+        await this.sendEvent(event)));
+      const successCount = resultArray.filter(r => r === true).length;
+      return {
+        totalEvents: events.length,
+        success: successCount,
+        failed: events.length - successCount,
+      }
+    } catch (err) {
+      logger.error(`Error occurred while metering all events : `, err);
+    }
+  }
+  /* jshint ignore:end */
+
+  static isServicePlanExcluded(options) {
+    const serviceId = _.get(options, 'service.service_guid');
+    const planId = _.get(options, 'service.plan_guid');
+    logger.info(`Checking if service ${serviceId}, plan: ${planId} is excluded`);
+    const plan = catalog.getPlan(planId);
+    return !plan.metered;
+  }
+
+  static enrichEvent(options) {
+    // Add region , service name and plan sku name of the event
+    const serviceId = _.get(options, 'service.service_guid');
+    const planId = _.get(options, 'service.plan_guid');
+    logger.info(`Enriching the metering event ${serviceId}, plan: ${planId}`);
+    options.service.id = catalog.getServiceName(serviceId);
+    options.service.plan = catalog.getPlanSKUFromPlanGUID(serviceId, planId);
+    options.service = _.omit(options.service, ['service_guid', 'plan_guid']);
+    options.consumer.region = config.metering.region;
+    return options;
+  }
+
+  /* jshint ignore:start */
+  static async sendEvent(event) {
+    try {
+      logger.debug('Metering Event details before enriching:', event);
+      if (this.isServicePlanExcluded(event.spec.options) === true) {
+        await this.updateMeterState(CONST.METER_STATE.EXCLUDED, event);
+        return true;
+      }
+      const enrichedUsageDoc = await this.enrichEvent(_.get(event.spec, 'options'));
+      logger.info('Sending enriched document:', enrichedUsageDoc);
+      const validEvent = await maas.client.sendUsageRecord({
+        usage: [enrichedUsageDoc]
+      });
+      if (validEvent !== undefined) {
+        this.updateMeterState(CONST.METER_STATE.METERED, event);
+      }
+      return true;
+    } catch (err) {
+      logger.error('Error occured while metering:', err);
+      await MeterInstanceJob._logMeteringEvent(err, event);
+      await this.updateMeterState(CONST.METER_STATE.FAILED, event, err);
+      return false;
+    }
+  }
+  /* jshint ignore:end */
+
+  static _logMeteringEvent(err, event) {
```

Comment:
shouldnt this method have asynch in it since its returning a promise? or asynch is to be added only to those methods which have await stmts in them and not returning a promise?

Reply:
We need `async` in definition only if we are awaiting for a promise inside the function.
Here we don't want to wait for the promise to complete before returning it